src/storage-controller/src/lib.rs

```diff
@@ -1886,16 +1886,17 @@ where
             instance.rehydrate_failed_replicas();
         }
 
-        let mut updated_frontiers = None;
+        let mut updated_frontiers = BTreeMap::new();
 
         // Take the allocation so that we can call mut receiver functions in the loop. We put it
         // back at the end of the loop.
         let mut stashed_responses = std::mem::take(&mut self.stashed_responses);
         for resp in stashed_responses.drain(..) {
             match resp {
-                StorageResponse::FrontierUppers(updates) => {
-                    self.update_write_frontiers(&updates);
-                    updated_frontiers = Some(Response::FrontierUpdates(updates));
+                StorageResponse::FrontierUpper(id, upper) => {
+                    let update = (id, upper);
+                    self.update_write_frontiers(std::slice::from_ref(&update));
```

Comment:
Looks like `Controller::update_write_frontiers` is now only called with a single update everywhere, so you could rename it to `update_write_frontier` and remove its loop.